### PR TITLE
Bulk edit

### PR DESCRIPTION
**Pull request by eddyonboard**

### CODE REVIEW

File: resources/assets/js/include/mass.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    $(document).bind('massforminit', function( e )
+    {
+        let target = $(e.target);
+        target = target.find('.edit-resources');
+
+        target.find('input.bulk-control').on('change', function(e){
+            if($(this).prop('checked')){
+                target.find('[name="resource['+$(this).attr('data-target')+']"]').prop("disabled", false);
+            } else {
+                target.find('[name="resource['+$(this).attr('data-target')+']"]').prop("disabled", true);
```
Comment (exabyssus) [inline resources/assets/js/include/mass.js]: Only thing that changes is true/false. Please use var and single line

File: resources/assets/js/include/mass.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        $(this).attr('href', getUrlwithIds)
+    }
+
+    function getUrlwithIds(){
+        let ids = $( "input.mass-row:checked" ).serializeArray();
+        return url + ( url.indexOf('?') >= 0 ? '&' : '?' ) + $.param(ids);
+    }
+
+    function allChecked() {
+        $( "input.mass-row").prop("checked", this.checked);
```
Comment (exabyssus) [inline resources/assets/js/include/mass.js]: Don't use "input", just better class selector.
Move to ar variable.

let bulkEditRow = $('.js-bulk-edit-row-checkbox');
And use it for operations.

bulkEditRow.filter(":checked")  and so on

File: src/Admin/Form/MassFormBuilder.php [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        ] );
+
+        $form->append( csrf_field() );
+
+        $form->append( Html::input()->setName( '_method' )->setType( 'hidden' )->setValue( 'POST' ) );
+
+        return $form;
+    }
+
+    protected function footer()
```
Comment (exabyssus) [inline src/Admin/Form/MassFormBuilder.php]: Add PHPDoc

File: src/Admin/Form/MassFormBuilder.php [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+
+    /**
+     * @return Content
+     */
+    public function render()
+    {
+        $content = Html::div()->addClass( 'body' );
+
+        //$this->addCheckboxesToEachInput();
```
Comment (exabyssus) [inline src/Admin/Form/MassFormBuilder.php]: Don't commit commented code

File: src/Admin/Grid.php [hunk 5]
```diff
@@ -211,12 +211,17 @@ public function getRows()
...
      * @return Column
      */
-    public function column( $name = null, $label = null )
+    public function column( $name = null, $label = null, $prepend = false )
     {
         $column = new Column( $name, $label );
         $column->setGrid( $this );
 
-        $this->columns->push( $column );
+        if($prepend) {
```
Comment (exabyssus) [inline src/Admin/Grid.php]: Why add this new functionality? If not necessary. 

Flags are not a good way to do this. In this case you should have used three methods - column() (for backwards compability), appendColumn(), prependColumn()

File: src/Admin/Traits/Bulk.php [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        return $form;
+    }
+
+    /**
+     * @param Model $model
+     * @param bool $ids
+     * @param bool $withChecks
+     * @return Form
+     */
+    protected function buildMassForm(Model $model, $ids = false, $withChecks = true)
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: Try not to use Flags and optional parameters.

File: src/Admin/Traits/Bulk.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        $form->setRenderer(new MassFormBuilder($form));
+        if($ids){
+            $form->title(trans('arbory::resources.mass_form_title', ['count' => count($ids)]));
+            $form->setFields(function (Form\FieldSet $fieldSet) use ($ids) {
+                $fieldSet->hidden('ids')->setValue(implode(',', $ids));
+            });
+        }
+        $this->massForm($form);
+
+        if($withChecks) {
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: This seems unnecessary functionality

File: src/Admin/Traits/Bulk.php [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+     * @return \Illuminate\Http\JsonResponse|\Illuminate\Http\RedirectResponse|\Illuminate\Routing\Redirector
+     */
+    public function massUpdate(Request $request)
+    {
+        $form = $this->buildMassForm($this->resource(), false, true);
+
+        $form->validate();
+
+        if ($request->ajax()) {
+            return response()->json(['ok']);
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: What is this? For validation?

File: src/Admin/Traits/Bulk.php [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        $input = $request->all('resource.ids');
+        $ids = isset($input['resource']['ids']) ? $input['resource']['ids'] : '';
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: Arr::get maybe?

File: src/Admin/Traits/Bulk.php [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        };
+
+    }
+
+    /**
+     * @param $form
+     */
+    protected function preprocessMassUpdate(Form $form)
+    {
+        $request = request();
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: Unecessary

File: src/Admin/Traits/Bulk.php [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        $clonedFields = clone $originalFields;
+
+        $items = $originalFields->getFields();
+
+        //count when iput added only, cache adjunctions in collection
+        $counter = 0;
+
+        foreach($clonedFields->getIterator() as $key => $field) {
+            $type = $field->getFieldTypeName();
+            if($type != 'type-hidden'){
```
Comment (exabyssus) [inline src/Admin/Traits/Bulk.php]: This is bad design. 
You should get "bulk edit fields" in another method. 
Use const as `public const BULK_UPDATE_FIELD_TYPES = ['select', 'input', 'multiselect']`;
Or have this propory on each Field class

File: resources/assets/js/include/mass.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        bulkEditRows.on('change', modifyUrl);
+        bulkEditHeader.on('change', allChecked);
+    }
+
+    /**
+     * Init grid events and try to init bulk form
+     */
+    $('body').on('contentloaded', function (e, event_params) {
+
+        if (bulkGrid.length) {
```
Comment (rerzy) [inline resources/assets/js/include/mass.js]: This should be part of bulkforminit event

File: resources/views/dialogs/form_mass_empty.blade.php [hunk 13]
```diff
@@ -0,0 +1,22 @@
+@extends('arbory::dialogs.base',['class' => 'mass-empty confirm'])
+
+@section('dialog')
+<header>
+    <h1>@lang('Empty selection')</h1>
```
Comment (rerzy) [inline resources/views/dialogs/form_mass_empty.blade.php]: Translations should be using dot notation like the rest of the arbory

**Pull request merged**